Source/CombatExtended/Utilities/Tracking/ThingsTrackingModel.cs

```diff
@@ -0,0 +1,210 @@
+
+        private HashSet<Thing> things = new HashSet<Thing>();
+        private ThingPositionInfo[] sortedThings = new ThingPositionInfo[100];
+
+        public ThingsTrackingModel(ThingDef def, Map map, ThingsTracker parent)
+        {
+            this.def = def;
+            this.map = map;
+            this.parent = parent;
+        }
+
+        public void Register(Thing thing)
+        {
+            if (indexByThing.ContainsKey(thing))
+                return;
+            if (things.Count + 1 >= sortedThings.Length)
+            {
+                ThingPositionInfo[] temp = new ThingPositionInfo[sortedThings.Length * 2];
+                for (int i = 0; i < sortedThings.Length; i++)
+                    temp[i] = sortedThings[i];
+                sortedThings = temp;
+            }
+            if (things.Count == 0)
+            {
+                sortedThings[things.Count] = new ThingPositionInfo(thing);
+                indexByThing[thing] = 0;
+                things.Add(thing);
+                return;
+            }
+            else
+            {
+                sortedThings[things.Count] = new ThingPositionInfo(thing);
+                indexByThing[thing] = things.Count - 1;
+                things.Add(thing);
+
+                int i = things.Count - 1;
+                while (i - 1 >= 0 && sortedThings[i] < sortedThings[i - 1])
+                {
+                    Swap<ThingPositionInfo>(i - 1, i, sortedThings);
+                    indexByThing[sortedThings[i].thing] = i;
+                    indexByThing[sortedThings[i - 1].thing] = i - 1;
+                    i--;
+                }
+            }
+        }
+
+        public void Remove(Thing thing)
+        {
+            if (!indexByThing.ContainsKey(thing))
+                return;
+            int index = indexByThing[thing];
+            int i;
+
+            for (i = index + 1; i < things.Count; i++)
+            {
+                indexByThing[sortedThings[i].thing] = i - 1;
+                sortedThings[i - 1] = sortedThings[i];
+            }
+            indexByThing.Remove(thing);
+            things.Remove(thing);
+        }
+
+        public void Notify_ThingPositionChanged(Thing thing)
+        {
+            int i;
+            int index = indexByThing[thing];
+
+            i = index;
+            while (i + 1 < things.Count && sortedThings[i] > sortedThings[i + 1])
+            {
+                Swap<ThingPositionInfo>(i + 1, i, sortedThings);
+                indexByThing[sortedThings[i].thing] = i;
+                indexByThing[sortedThings[i + 1].thing] = i + 1;
+                i++;
+            }
+            i = index;
+            while (i - 1 >= 0 && sortedThings[i] < sortedThings[i - 1])
+            {
+                Swap<ThingPositionInfo>(i - 1, i, sortedThings);
+                indexByThing[sortedThings[i].thing] = i;
+                indexByThing[sortedThings[i - 1].thing] = i - 1;
+                i--;
+            }
+        }
```

Comment:
This seems a bit scary to me, every time a thing's position changes you do an `O(n)` operation

Reply:
Nope, since things can only move so fast. Since things move with limited speed it's an average 1 swap operation.